crates/matrix-sdk/src/send_queue/mod.rs

```diff
@@ -580,7 +601,57 @@ impl RoomSendQueue {
                 continue;
             };
 
-            match Self::handle_request(&room, queued_request, cancel_upload_rx).await {
+            let default_index = 0; // Before MSC4274 only a single file (and thumbnail) could be sent per event.
+            let index = {
+                cfg_if::cfg_if! {
+                    if #[cfg(feature = "unstable-msc4274")] {
+                        match &queued_request.kind {
+                            QueuedRequestKind::MediaUpload { accumulated, .. } => accumulated.len() as u64,
+                            _ => default_index,
+                        }
+                    } else {
+                        default_index
+                    }
+                }
+            };
+
+            // Prepare to watch and communicate the request progress.
+            let progress = if let Some(related_to) = related_txn_id.as_ref() {
+                let progress: SharedObservable<TransmissionProgress> = Default::default();
+                let mut subscriber = progress.subscribe();
+
+                let updates = updates.clone();
+                let related_to = related_to.clone();
+
+                // Watch and communicate the progress on a detached background task. Once
+                // the progress observable is dropped, next() will return None and the
+                // task will end.
+                spawn(async move {
```

Comment:
Does it correctly terminate when the upload is cancelled? (needs a test!)

Reply:
I moved this comment to https://github.com/matrix-org/matrix-rust-sdk/pull/5008/files#r2190753736 as this position is now outdated.